### ISSUE TITLE
Document expected return values of unsupported file types

### ISSUE THREAD

**Issue by jessrosenfield** [sindresorhus/file-type#377]
Because these following file types are not supported, it would be helpful to users to know what ext/mime `file-type` will return for them (eg: `application/x-msi`)

**Comment by sindresorhus** [comment]
From the readme:

> Or undefined when there is no match.

**Comment by jessrosenfield** [comment]
@sindresorhus  The problem is unsupported file types don't always return undefined. As I mentioned above, office <=2004 files return `application/x-msi`.

**Comment by sindresorhus** [comment]
I guess we could document the known ones, yes.

**Comment by jessrosenfield** [comment]
I'm happy to make a PR for a change to the readme. I'm just curious, are there any other known return values for unsupported file types besides `{ext: "msi", mime: "application/x-msi"}` and `undefined`?

**Comment by Borewit** [comment]
Issue #290. The ASF media container is now always mapped to .wma or .wmv. Which is thw wrong way around. .wma & .wmv are more specific then ASF.

WebM audio is detected as WebM video, issue #271.

There are more types which are detected as more generic types as the original file may have been.

**Comment by niftylettuce** [comment]
Unsupported file types should return `undefined`, they shouldn't return a value in my opinion.

**Comment by niftylettuce** [comment]
@jessrosenfield Can you make a PR to the README? 🎉 👍

**Comment by niftylettuce** [comment]
@sindresorhus is there any way we can have it return undefined here instead or some other way of letting us know it's just an office file <= 2004? <https://github.com/sindresorhus/file-type/blob/3effaeed691131a00d00fca6fc6e695dac393d62/core.js#L1215-L1220>

**Comment by Borewit** [comment]
> @sindresorhus is there any way we can have it return undefined here instead or some other way of letting us know it's just an office file <= 2004?

The answer to that is already covered in the the README @niftylettuce. But I am more then happy to review your disproving PR.

**Comment by niftylettuce** [comment]
@Borewit it's not obvious what the answer is?  Is it modifying `FileType.extensions` to exclude `msi` and also similarly excluding it in `.mime`?

**Comment by niftylettuce** [comment]
As far I can tell there's only getters, no setters..

**Comment by Borewit** [comment]
That [code fragment](https://github.com/sindresorhus/file-type/blob/3effaeed691131a00d00fca6fc6e695dac393d62/core.js#L1215-L1220) essentially recognizes an [[MS-CFB]: Compound File Binary File Format](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-cfb/05060311-bfce-4b12-874d-71fd4ce63aea), which could be an MSI file. But it could be a Word 97-2003 `.doc` file as well or another file using this Microsoft "general purpose" archive file format. I think you have a fair point there @niftylettuce.

**Comment by niftylettuce** [comment]
This is what I have to do right now to workaround this in my work with @spamscanner:

```js
const fileType = await FileType.fromBuffer(attachment.content);
// TODO: msi extensions are currently allowed since `undefined` is not yet returned
// <https://github.com/sindresorhus/file-type/issues/377>
if (
  fileType &&
  fileType.ext &&
  fileType.ext !== 'msi' &&
  fileType.mime &&
  fileType.mime !== 'application/x-msi' &&
  EXECUTABLES.includes(fileType.ext)
)
  messages.push(
    `Attachment's "magic number" indicated it was a dangerous executable with a ".${fileType.ext}" extension.`
  );
```

**Comment by niftylettuce** [comment]
Obviously this isn't ideal, since "msi" files could actually be executables and malicious (see previous snippet).

**Comment by sindresorhus** [comment]
Well, one solution is for you to just detect `.doc` yourself and return `undefined` before running it through `file-type`.

We do want to support custom matchers https://github.com/sindresorhus/file-type/issues/340.

**Comment by niftylettuce** [comment]
This code I'm doing only allows me access to a Buffer of file content.  It's an extreme edge case but a vulnerability.

```js
        if (isBuffer(attachment.content)) {
          try {
            const fileType = await FileType.fromBuffer(attachment.content);
            // TODO: msi extensions are currently allowed since `undefined` is not yet returned
            // <https://github.com/sindresorhus/file-type/issues/377>
            if (
              fileType &&
              fileType.ext &&
              fileType.ext !== 'msi' &&
              fileType.mime &&
              fileType.mime !== 'application/x-msi' &&
              EXECUTABLES.includes(fileType.ext)
            )
              messages.push(
                `Attachment's "magic number" indicated it was a dangerous executable with a ".${fileType.ext}" extension.`
              );
          } catch (err) {
            this.config.logger.error(err);
          }
        }

        if (isSANB(attachment.filename)) {
          // run punycode on file name as well since file names can be spoofed
          const filename = punycode.toASCII(
            punycode.toUnicode(attachment.filename.split('?')[0])
          );
          const ext = fileExtension(filename);
          if (ext && EXECUTABLES.includes(ext))
            messages.push(
              `Attachment's file name indicated it was a dangerous executable with a ".${ext}" extension.`
            );
        }

        if (isSANB(attachment.contentType)) {
          const ext = mime.extension(attachment.contentType);
          if (isSANB(ext) && EXECUTABLES.includes(ext))
            messages.push(
              `Attachment's Content-Type was a dangerous executable with a ".${ext}" extension.`
            );
        }
```

I still at least check for `msi` extension as shown in the code.

**Comment by niftylettuce** [comment]
Can this be hooked into file-type by default @sindresorhus ?

**Comment by niftylettuce** [comment]
Regardless, thank you so much @Borewit 🙇 👍

**Comment by niftylettuce** [comment]
@Borewit something odd I noticed, it doesn't always work.  For example, I created a file in LIbreOffice and saved as Word 97-2003 (and it did not even get an original file type using `file-type`.  Attached are two files "word-97-2003.doc" (LibreOffice created) and "sample.doc" (a Word file I found online- this one DOES work with the detection example you gave).


[Archive.zip](https://github.com/sindresorhus/file-type/files/5114805/Archive.zip)

**Comment by niftylettuce** [comment]
Ah, I see this note in the README:

> .doc - Too old and difficult to parse.

Nevermind.

**Comment by sindresorhus** [comment]
> Can this be hooked into file-type by default @sindresorhus ?

No, I don't want to deal with this file format. https://github.com/sindresorhus/file-type/issues/340 is a better way forward.

**Comment by niftylettuce** [comment]
Just a heads up @Borewit there is a core breaking issue with that package, see <https://github.com/ifedoroff/compound-file-js/pull/2>.

**Comment by niftylettuce** [comment]
To anyone reading this, there are core bugs in the original `compound-binary-file-js` package that are not yet merged/released.  I've fixed them and published `v1.0.11` of `@spamscanner/compound-binary-file-js` in the interim per my comments at <https://github.com/ifedoroff/compound-file-js/pull/2>.  It looks like @Borewit submitted a PR that resolves this altogether though. 👍
